File: matlab/gpCovGrads.m
function [gK_uu, gK_uf, g_Lambda, g_Lambda2] = gpCovGrads(model, Y)

% GPCOVGRADS Sparse objective function gradients wrt Covariance functions for inducing variables.
%
% [gK_uu, gK_uf, g_Lambda, g_Lambda2] = gpCovGrads(model, Y)
%

% Copyright (c) 2006 Neil D. Lawrence
% gpCovGrads.m version 1.3



switch model.approx
 case 'dtc'
  % Deterministic training conditional.
  if ~isfield(model, 'isSpherical') | model.isSpherical
    E = model.K_uf*Y;
    EET = E*E';
    AinvEET = model.Ainv*EET;
    AinvEETAinv = AinvEET*model.Ainv;
    gK_uu = 0.5*(model.d*(model.invK_uu-(1/model.beta)*model.Ainv) ...
                 - AinvEETAinv);
    
    AinvK_uf = model.Ainv*model.K_uf;
    gK_uf = -model.d*AinvK_uf-model.beta*(AinvEET*AinvK_uf-(model.Ainv*E*Y'));
    
    g_Lambda2 = 0.5*(model.d*((model.N-model.k)/model.beta ...
                              +sum(sum(model.Ainv.*model.K_uu))/(model.beta*model.beta))...
                     +sum(sum(AinvEETAinv.*model.K_uu))/model.beta ...
                     +(trace(AinvEET)-sum(sum(Y.*Y))));
    for k = 1:size(model.m, 2)
     if model.KLCorrectionTerm
         g_Lambda2 = g_Lambda2 - 0.5.*sum(model.KLVariance(:,k)-model.m(:,k).*model.m(:,k));       
     end
    end
    
    fhandle = str2func([model.betaTransform 'Transform']);
    g_Lambda2 = g_Lambda2*fhandle(model.beta, 'gradfact');
    g_Lambda = [];
  else
    gK_uu = zeros(model.k, model.k);
    gK_uf = zeros(model.k, model.N);
    g_Lambda2 = 0;
    for i = 1:model.d
      ind = gpDataIndices(model, i);
      e = model.K_uf(:, ind)*Y(ind, i);
      Ainve = model.Ainv{i}*e;
      AinveeT = Ainve*e';      
      AinveeTAinv = Ainve*Ainve';
      gK_uu = gK_uu+0.5*((model.invK_uu-(1/model.beta)*model.Ainv{i}) ...
                         - AinveeTAinv);
      
      AinvK_uf = model.Ainv{i}*model.K_uf(:, ind);
      gK_uf(:, ind) = gK_uf(:, ind) - AinvK_uf...
          -model.beta*(AinveeT*AinvK_uf-(Ainve*Y(ind, i)'));
      
      g_Lambda2 = g_Lambda2 ...
          + 0.5*(((model.N-model.k)/model.beta ...
                  +sum(sum(model.Ainv{i}.*model.K_uu))/(model.beta*model.beta))...
                 +sum(sum(AinveeTAinv.*model.K_uu))/model.beta ...
                 +(trace(AinveeT)-sum(sum(Y(ind, i).*Y(ind, i)))));
    end
    for k = 1:size(model.m, 2)
    if model.KLCorrectionTerm
        g_Lambda2 = g_Lambda2 - 0.5.*sum(model.KLVariance(:,k)-model.m(:,k).*model.m(:,k));       
    end
    end
    fhandle = str2func([model.betaTransform 'Transform']);
    g_Lambda2 = g_Lambda2*fhandle(model.beta, 'gradfact');
    g_Lambda = [];
  end
 case 'fitc'
  % Fully independent training conditonal.
  if ~isfield(model, 'isSpherical') | model.isSpherical
    K_ufDinvK_uf = model.K_uf*model.Dinv*model.K_uf';
    E = model.K_uf*model.Dinv*Y;
    EET = E*E';
    AinvEET = model.Ainv*EET;
    diagK_fuAinvEYT = sum(model.K_uf.*(model.Ainv*E*Y'), 1)';
    AinvEETAinv = AinvEET*model.Ainv;
    diagK_ufdAinvplusAinvEETAinvK_fu = ...
        sum(model.K_uf.*((model.d*model.Ainv+AinvEETAinv)*model.K_uf), 1)';
    invK_uuK_uf = model.invK_uu*model.K_uf;
    invK_uuK_ufDinv = invK_uuK_uf*model.Dinv;
    diagYYT = sum(Y.*Y, 2);
    diagQ = -model.d*model.diagD + diagYYT ...
            + diagK_ufdAinvplusAinvEETAinvK_fu...
            -2*diagK_fuAinvEYT;
    gK_uu = 0.5*(model.d*model.invK_uu ...
                 - model.d*model.Ainv - AinvEETAinv ...
                 + invK_uuK_ufDinv*sparseDiag(diagQ)*invK_uuK_ufDinv');
    gK_uf = -invK_uuK_ufDinv*sparseDiag(diagQ)*model.Dinv ...      
            -model.d*model.Ainv*model.K_uf*model.Dinv ...
            -AinvEETAinv*model.K_uf*model.Dinv ...
            +model.Ainv*E*Y'*model.Dinv;
    g_Lambda = 0.5*diagQ.*1./(model.diagD.*model.diagD);
    g_Lambda2 = -sum(g_Lambda)/(model.beta*model.beta);
    for k = 1:size(model.m, 2)
    if model.KLCorrectionTerm
        g_Lambda2 = g_Lambda2 - 0.5.*sum(model.KLVariance(:,k)-model.m(:,k).*model.m(:,k));       
    end
    end
    
    fhandle = str2func([model.betaTransform 'Transform']);
    g_Lambda2 = g_Lambda2*fhandle(model.beta, 'gradfact');
  else
    gK_uu = zeros(model.k, model.k);
    gK_uf = zeros(model.k, model.N);
    g_Lambda = zeros(model.N, 1);
    g_Lambda2 = 0;
    for i = 1:model.d
      ind = gpDataIndices(model, i);
      K_ufDinvK_uf = model.K_uf(:, ind)*model.Dinv{i}*model.K_uf(:, ...
                                                        ind)';
      e = model.K_uf(:, ind)*model.Dinv{i}*Y(ind, i);
      Ainve = model.Ainv{i}*e;
      AinveeTAinv = Ainve*Ainve';
      diagK_fuAinveyT = sum(model.K_uf(:, ind).*(Ainve*Y(ind,i)'), 1)';
      diagK_ufdAinvplusAinveeTAinvK_fu = ...
          sum(model.K_uf(:, ind).*((model.Ainv{i}+AinveeTAinv)*model.K_uf(:, ...
                                                        ind)), 1)';
      invK_uuK_uf = model.invK_uu*model.K_uf(:, ind);
      invK_uuK_ufDinv = invK_uuK_uf*model.Dinv{i};
      diagyyT = Y(ind, i).*Y(ind, i);
      diagQ = -model.diagD{i} + diagyyT ...
              + diagK_ufdAinvplusAinveeTAinvK_fu...
              -2*diagK_fuAinveyT;
      gK_uu = gK_uu ...
              +0.5*(model.invK_uu ...
                    - model.Ainv{i} - AinveeTAinv ...
                    + invK_uuK_ufDinv*sparseDiag(diagQ)*invK_uuK_ufDinv');
      gK_uf(:, ind) = gK_uf(:, ind) ...
              -invK_uuK_ufDinv*sparseDiag(diagQ)*model.Dinv{i} ...      
              -model.Ainv{i}*model.K_uf(:, ind)*model.Dinv{i} ...
              -AinveeTAinv*model.K_uf(:, ind)*model.Dinv{i} ...
              +Ainve*Y(ind, i)'*model.Dinv{i};
      g_Lambda(ind) = g_Lambda(ind) ...
          + 0.5*diagQ.*1./(model.diagD{i}.*model.diagD{i});
    end
    g_Lambda2 = g_Lambda2 - sum(g_Lambda)/(model.beta*model.beta);
    for k = 1:size(model.m, 2)
    if model.KLCorrectionTerm
        g_Lambda2 = g_Lambda2 - 0.5.*sum(model.KLVariance(:,k)-model.m(:,k).*model.m(:,k));       
    end
    end
    
    fhandle = str2func([model.betaTransform 'Transform']);
    g_Lambda2 = g_Lambda2*fhandle(model.beta, 'gradfact');
  end
 
 case 'pitc' 
  % Partially independent training conditional.
  if ~isfield(model, 'isSpherical') | model.isSpherical
    E = zeros(model.k, model.d);
    startVal = 1;
    for i = 1:length(model.blockEnd)
      endVal = model.blockEnd(i);
      ind = startVal:endVal;
      E = E + model.K_uf(:, ind)*model.Dinv{i}*Y(ind, :);
      startVal = endVal + 1;
    end
    AinvE = model.Ainv*E;
    AinvEET = AinvE*E';
    AinvEETAinv = AinvEET*model.Ainv;
    startVal = 1;
    for i = 1:length(model.blockEnd)
      endVal = model.blockEnd(i);
      ind = startVal:endVal;
      K_fuAinvEYT = model.K_uf(:, ind)'*AinvE*Y(ind, :)';
      blockQ{i} = -model.d*model.D{i} + Y(ind, :)*Y(ind, :)' ...
          + model.K_uf(:, ind)'*(model.d*model.Ainv + AinvEETAinv)*model.K_uf(:, ind)...
          -K_fuAinvEYT - K_fuAinvEYT';
      startVal = endVal + 1;
    end
    gK_uu = model.d*model.invK_uu ...
            - model.d*model.Ainv - AinvEETAinv;
    g_Lambda2 = 0;
    gK_ufBase = -(model.d*model.Ainv + AinvEETAinv)*model.K_uf ...
        + AinvE*Y';
    
    startVal = 1;
    for i = 1:length(model.blockEnd)
      endVal = model.blockEnd(i);
      ind = startVal:endVal;
      invK_uuK_ufDinv = model.invK_uu*model.K_uf(:, ind)*model.Dinv{i};
      gK_uu = gK_uu + invK_uuK_ufDinv*blockQ{i}*invK_uuK_ufDinv';
      
      gK_uf(:, ind) = (gK_ufBase(:, ind) ...
                       -invK_uuK_ufDinv*blockQ{i})*model.Dinv{i};
      
      
      g_Lambda{i} = 0.5*model.Dinv{i}*blockQ{i}*model.Dinv{i};
      g_Lambda2 = g_Lambda2 - sum(diag((g_Lambda{i})))/(model.beta*model.beta);
      startVal = endVal + 1;
    end
    gK_uu = gK_uu*0.5;
    for k = 1:size(model.m, 2)
    if model.KLCorrectionTerm
        g_Lambda2 = g_Lambda2 - 0.5.*sum(model.KLVariance(:,k)-model.m(:,k).*model.m(:,k));       
    end
    end
    fhandle = str2func([model.betaTransform 'Transform']);
    g_Lambda2 = g_Lambda2*fhandle(model.beta, 'gradfact');
  else
    gK_uu = zeros(model.k, model.k);
    gK_uf = zeros(model.k, model.N);
    for i = 1:length(model.blockEnd)
      if i == 1
        indLen = model.blockEnd(1);
      else
        indLen = model.blockEnd(i) - model.blockEnd(i-1);
      end
      g_Lambda{i} = zeros(indLen, indLen);
    end
    g_Lambda2 = 0;
    for j = 1:model.d
      e = zeros(model.k, 1);
      for i = 1:length(model.blockEnd)
        ind = gpDataIndices(model, j, i);
        e = e + model.K_uf(:, ind)*model.Dinv{i, j}*Y(ind, j);
      end
      Ainve = model.Ainv{j}*e;
      AinveeT = Ainve*e';
      AinveeTAinv = AinveeT*model.Ainv{j};
      for i = 1:length(model.blockEnd)
        ind = gpDataIndices(model, j, i);
        K_fuAinveyT = model.K_uf(:, ind)'*Ainve*Y(ind, j)';
        blockQ{i} = -model.D{i, j} + Y(ind, j)*Y(ind, j)' ...
            + model.K_uf(:, ind)'*(model.Ainv{j} + AinveeTAinv)*model.K_uf(:, ind)...
            -K_fuAinveyT - K_fuAinveyT';
      end
      gK_uu = gK_uu + model.invK_uu ...
            - model.Ainv{j} - AinveeTAinv;
      gK_ufBase = -(model.d*model.Ainv{i} + AinveeTAinv)*model.K_uf ...
        + Ainve*Y(:, j)';
    
      for i = 1:length(model.blockEnd)
        ind = gpDataIndices(model, j, i);
        invK_uuK_ufDinv = model.invK_uu*model.K_uf(:, ind)*model.Dinv{i,j};
        gK_uu = gK_uu + invK_uuK_ufDinv*blockQ{i}*invK_uuK_ufDinv';
      
        gK_uf(:, ind) = (gK_ufBase(:, ind) ...
                         -invK_uuK_ufDinv*blockQ{i})*model.Dinv{i,j};
      
        if i == 1
          localInd = ind;
        else
          localInd = ind - (model.blockEnd(i-1));
        end
        g_Lambda{i}(localInd, localInd) = g_Lambda{i}(localInd, localInd) ...
            + 0.5*model.Dinv{i,j}*blockQ{i}*model.Dinv{i,j};
      end
    end
    for i = 1:length(model.blockEnd)
      g_Lambda2 = g_Lambda2 - sum(diag((g_Lambda{i})))/(model.beta*model.beta);
    end
    gK_uu = gK_uu*0.5;
    for k = 1:size(model.m, 2)
    if model.KLCorrectionTerm
        g_Lambda2 = g_Lambda2 - 0.5.*sum(model.KLVariance(:,k)-model.m(:,k).*model.m(:,k));       
    end
    end
    fhandle = str2func([model.betaTransform 'Transform']);
    g_Lambda2 = g_Lambda2*fhandle(model.beta, 'gradfact');
  end
 otherwise
  error('Unknown approximation type');
end

File: R/gpCovGrads.R
gpCovGrads <- function(model, M) {

  if (model$approx %in% list('dtc', 'dtcvar')) {
    ## Deterministic training conditional.
    if (model$approx == 'dtcvar')
      dtcvar = TRUE
    else
      dtcvar = FALSE

    if (!'isSpherical' %in% names(model) || model$isSpherical) {
      E = model$K_uf%*%M
      EET = E%*%t(E)
      AinvEET = model$Ainv%*%EET
      AinvEETAinv = AinvEET%*%model$Ainv
      gK_uu = 0.5*(model$d*(model$invK_uu-(1/model$beta)*model$Ainv) - AinvEETAinv)
      if (dtcvar) {
	K_uuInvK_uf = model$invK_uu%*%model$K_uf
	gK_uu = gK_uu - 0.5*model$d*model$beta*(K_uuInvK_uf%*%t(K_uuInvK_uf))
      }
      AinvK_uf = model$Ainv%*%model$K_uf
      gK_uf = -model$d*AinvK_uf - model$beta*(AinvEET%*%AinvK_uf - (model$Ainv%*%E%*%t(M)))
      if (dtcvar)
	gK_uf = gK_uf + model$d*model$beta*K_uuInvK_uf

      gBeta = 0.5*(model$d*((model$N-model$k)/model$beta
		      +sum(model$Ainv * model$K_uu)/(model$beta*model$beta))
		      +sum(AinvEETAinv * model$K_uu)/model$beta
		      +(sum(diag(AinvEET))-sum(M * M)))
      if (dtcvar)
	gBeta = gBeta -0.5*model$d*sum(model$diagD)/model$beta

      fhandle = get(model$betaTransform$func, mode="function")
      gBeta = gBeta*fhandle(model$beta, 'gradfact')
      if (dtcvar)
	g_Lambda = matrix(-0.5*model$beta*model$d, 1, model$N)
      else
	g_Lambda = matrix()
    } else {
      gK_uu = matrix(0, model$k, model$k)
      gK_uf = matrix(0, model$k, model$N)
      gBeta = 0
      for (i in 1:model$d) {
	ind = gpDataIndices(model, i)
	e = model$K_uf[, ind,drop=F]%*%M[ind, i,drop=F]
	Ainve = model$Ainv[[i]]%*%e
	AinveeT = Ainve%*%t(e)      
	AinveeTAinv = Ainve%*%t(Ainve)
	gK_uu = gK_uu + 0.5*((model$invK_uu - (1/model$beta)*model$Ainv[[i]]) - AinveeTAinv)

	AinvK_uf = model$Ainv[[i]]%*%model$K_uf[, ind,drop=F]
	gK_uf[, ind] = gK_uf[, ind,drop=F] - AinvK_uf
	    - model$beta*(AinveeT%*%AinvK_uf - (Ainve%*%t(M[ind, i, drop=F])))

	gBeta = gBeta + 0.5*(((model$N - model$k)/model$beta
		  +sum(model$Ainv[[i]] * model$K_uu)/(model$beta*model$beta))
		  +sum(AinveeTAinv * model$K_uu)/model$beta
		  +(sum(diag(AinveeT))-sum(M[ind, i,drop=F] * M[ind, i,drop=F])))
      }
      fhandle = get(model$betaTransform$func, mode="function")
      gBeta = gBeta*fhandle(model$beta, 'gradfact')
      g_Lambda = matrix()
    }
  } else if (model$approx == 'fitc') {
    ## Fully independent training conditonal.
    if (!'isSpherical' %in% names(model) || model$isSpherical) {
      E = model$K_uf %*% model$Dinv %*% M
      EET = E%*%t(E)
      AinvE = model$Ainv%*%E
      diagK_fuAinvEMT = t(colSums(model$K_uf * (model$Ainv%*%E%*%t(M))))
      AinvEETAinv = AinvE%*%t(AinvE)
      diagK_ufdAinvplusAinvEETAinvK_fu =
	t(colSums(model$K_uf * ((model$d*model$Ainv + model$beta*AinvEETAinv)%*%model$K_uf)))
      invK_uuK_uf = model$invK_uu%*%model$K_uf
      if (TRUE)
	invK_uuK_ufDinv = invK_uuK_uf%*%model$Dinv
      else
	invK_uuK_ufDinv = solve(t(model$L), model$V)

      diagMMT = rowSums(M * M)
      diagQ = - model$d*model$diagD + model$beta*diagMMT
	      + diagK_ufdAinvplusAinvEETAinvK_fu - 2*model$beta*diagK_fuAinvEMT
      gK_uu = 0.5*(model$d*(model$invK_uu - model$Ainv/model$beta) - AinvEETAinv
	      + model$beta*invK_uuK_ufDinv%*%diag.spam(drop(diagQ))%*%t(invK_uuK_ufDinv))
      gK_uf = -model$beta*invK_uuK_ufDinv%*%diag.spam(drop(diagQ))%*%model$Dinv
	      -model$d*model$Ainv%*%model$K_uf%*%model$Dinv
	      -model$beta*AinvEETAinv%*%model$K_uf%*%model$Dinv
	      +model$beta*model$Ainv%*%E%*%t(M)%*%model$Dinv
      g_Lambda = (0.5*diagQ*model$beta) / (model$diagD * model$diagD)
      gBeta = -sum(g_Lambda)/(model$beta*model$beta)
      fhandle = get(model$betaTransform$func, mode="function")
      gBeta = gBeta*fhandle(model$beta, 'gradfact')
    } else {
      gK_uu = matrix(0, model$k, model$k)
      gK_uf = matrix(0, model$k, model$N)
      g_Lambda = matrix(0, model$N, 1)
      gBeta = 0
      for (i in 1:model$d) {
	ind = gpDataIndices(model, i)
	K_ufDinvK_uf = model$K_uf[, ind,drop=F]%*%model$Dinv[[i]]%*%t(model$K_uf[, ind,drop=F])
	e = model$K_uf[, ind,drop=F]%*%model$Dinv[[i]]%*%M[ind, i,drop=F]
	Ainve = model$Ainv[[i]]%*%e
	AinveeTAinv = Ainve%*%t(Ainve)
	diagK_fuAinveyT = t(colSums(model$K_uf[, ind,drop=F] * (Ainve%*%t(M[ind, i,drop=F]))))
	diagK_ufdAinvplusAinveeTAinvK_fu =
	    t(colSums(model$K_uf[,ind,drop=F] * ((model$Ainv[[i]]
	      + model$beta*AinveeTAinv)%*%model$K_uf[, ind,drop=F])))
	invK_uuK_uf = model$invK_uu%*%model$K_uf[, ind,drop=F]
	invK_uuK_ufDinv = invK_uuK_uf%*%model$Dinv[[i]]
	diagyyT = M[ind, i,drop=F] * M[ind, i,drop=F]
	diagQ = -model$diagD[[i]] + model$beta*diagyyT
		+ diagK_ufdAinvplusAinveeTAinvK_fu - 2*model$beta*diagK_fuAinveyT
	gK_uu = gK_uu + 0.5*(model$invK_uu - model$Ainv[[i]]/model$beta - AinveeTAinv
	  + model$beta*invK_uuK_ufDinv%*%diag.spam(drop(diagQ))%*%t(invK_uuK_ufDinv))
	gK_uf[, ind] = gK_uf[, ind,drop=F]
		-model$beta*invK_uuK_ufDinv%*%diag.spam(drop(diagQ))%*%model$Dinv[[i]]
		-model$Ainv[[i]]%*%model$K_uf[, ind,drop=F]%*%model$Dinv[[i]]
		-model$beta*AinveeTAinv%*%model$K_uf[, ind,drop=F]%*%model$Dinv[[i]]
		+model$beta*Ainve%*%t(M[ind, i,drop=F])%*%model$Dinv[[i]]
	g_Lambda[ind] = g_Lambda[ind]
	    + 0.5*model$beta*diagQ / (model$diagD[[i]] * model$diagD[[i]])
      }
      gBeta = gBeta - sum(g_Lambda)/(model$beta*model$beta)
      fhandle = get(model$betaTransform$func, mode="function")
      gBeta = gBeta*fhandle(model$beta, 'gradfact')
    }
  } else if (model$approx == 'pitc') {
    ## Partially independent training conditional.
    if (!'isSpherical' %in% names(model) || model$isSpherical) {
      E = matrix(0, model$k, model$d)
      for (i in 1:length(model$blockEnd)) {
	ind = gpBlockIndices(model, i)
	E = E + model$K_uf[, ind,drop=F]%*%model$Dinv[[i]]%*%M[ind, ,drop=F]
      }
      AinvE = model$Ainv%*%E
      AinvEET = AinvE%*%t(E)
      AinvEETAinv = AinvEET%*%model$Ainv
      blockQ = list()
      for (i in 1:length(model$blockEnd)) {
	ind = gpBlockIndices(model, i)
	K_fuAinvEMT = model$beta*t(model$K_uf[, ind,drop=F])%*%AinvE%*%t(M[ind, ,drop=F])
	blockQ[[i]] = -model$d*model$D[[i]] + model$beta*M[ind, ,drop=F]%*%t(M[ind, ,drop=F])
	    + t(model$K_uf[, ind,drop=F])%*%(model$d*model$Ainv + model$beta*AinvEETAinv)%*%
	    model$K_uf[, ind,drop=F] - K_fuAinvEMT - t(K_fuAinvEMT)
      }
      gK_uu = model$d*model$invK_uu - model$d*model$Ainv/model$beta - AinvEETAinv
      gBeta = 0
      gK_ufBase = -(model$d*model$Ainv + model$beta*AinvEETAinv)%*%model$K_uf
	  + model$beta*AinvE%*%t(M)

      g_Lambda = list()
      gK_uf = matrix(0, dim(gK_ufBase)[1], model$N)
      for (i in 1:length(model$blockEnd)) {
	ind = gpBlockIndices(model, i)
	invK_uuK_ufDinv = model$invK_uu%*%model$K_uf[, ind,drop=F]%*%model$Dinv[[i]]
	gK_uu = gK_uu + model$beta*invK_uuK_ufDinv%*%blockQ[[i]]%*%t(invK_uuK_ufDinv)

	gK_uf[, ind] = (gK_ufBase[,ind,drop=F]-model$beta*invK_uuK_ufDinv%*%blockQ[[i]])%*%model$Dinv[[i]]

	g_Lambda[[i]] = 0.5*model$Dinv[[i]]%*%blockQ[[i]]%*%model$Dinv[[i]]*model$beta
	gBeta = gBeta - sum(diag((g_Lambda[[i]]))) / (model$beta*model$beta)
      }
      gK_uu = gK_uu*0.5
      fhandle = get(model$betaTransform$func, mode="function")
      gBeta = gBeta*fhandle(model$beta, 'gradfact')
    } else {
      gK_uu = matrix(0, model$k, model$k)
      gK_uf = matrix(0, model$k, model$N)
      g_Lambda = list()
      for (i in 1:length(model$blockEnd)) {
	if (i == 1)
	  indLen = model$blockEnd[1]
	else
	  indLen = model$blockEnd[i] - model$blockEnd[i-1]

	g_Lambda[[i]] = matrix(0, indLen, indLen)
      }
      gBeta = 0
      for (j in 1:model$d) {
	e = matrix(0, model$k, 1)
	for (i in 1:length(model$blockEnd)) {
	  ind = gpDataIndices(model, j, i)
	  e = e + model$K_uf[, ind,drop=F]%*%model$Dinv[[i]][[j]]%*%M[ind, j,drop=F]
	}
	Ainve = model$Ainv[[j]]%*%e
	AinveeT = Ainve%*%t(e)
	AinveeTAinv = AinveeT%*%model$Ainv[[j]]
	blockQ = list()
	for (i in 1:length(model$blockEnd)) {
	  ind = gpDataIndices(model, j, i)
	  K_fuAinveyT = model$beta*t(model$K_uf[, ind,drop=F])%*%Ainve%*%t(M[ind, j,drop=F])
	  blockQ[[i]] = -model$D[[i]][[j]] + model$beta*M[ind, j,drop=F]%*%t(M[ind, j,drop=F])
	      + t(model$K_uf[, ind,drop=F])%*%(model$Ainv[[j]] + model$beta*AinveeTAinv)%*%
		model$K_uf[, ind,drop=F] - K_fuAinveyT - t(K_fuAinveyT)
	}
	gK_uu = gK_uu + model$invK_uu - model$Ainv[[j]]/model$beta - AinveeTAinv

	for (i in 1:length(model$blockEnd)) {
	  ind = gpDataIndices(model, j, i)
	  gK_ufBase = -(model$Ainv[[i]] + model$beta*AinveeTAinv)%*%model$K_uf[, ind,drop=F]
	      + model$beta*Ainve%*%t(M[ind, j,drop=F])
	  invK_uuK_ufDinv = model$invK_uu%*%model$K_uf[, ind,drop=F]%*%model$Dinv[[i]][[j]]
	  gK_uu = gK_uu + model$beta*invK_uuK_ufDinv%*%blockQ[[i]]%*%t(invK_uuK_ufDinv)

	  gK_uf[, ind] = gK_uf[, ind,drop=F] + (gK_ufBase
			  -model$beta*invK_uuK_ufDinv%*%blockQ[[i]])%*%model$Dinv[[i]][[j]]

	  if (i == 1)
	    localInd = ind
	  else
	    localInd = ind - (model$blockEnd[i-1])

	  g_Lambda[[i]][localInd, localInd] = g_Lambda[[i]][localInd, localInd]
	      + 0.5*model$Dinv[[i]][[j]]%*%blockQ[[i]]%*%model$Dinv[[i]][[j]]*model$beta
	}
      }

      for (i in 1:length(model$blockEnd))
	gBeta = gBeta - sum(diag((g_Lambda[[i]])))/(model$beta*model$beta)

      gK_uu = gK_uu*0.5
      fhandle = get(model$betaTransform$func, mode="function")
      gBeta = gBeta*fhandle(model$beta, 'gradfact')
    }
  } else
    stop("Unknown approximation type")

  return (list(gK_uu=gK_uu, gK_uf=gK_uf, g_Lambda=g_Lambda, gBeta=gBeta))
}
